File: Python/pathSum.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: pathSum.py
@Function: 路径总和 III 
@Link: https://leetcode-cn.com/problems/path-sum-iii/
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-07-31
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# 精简版，五行代码不解释
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        def dfs(root, sumlist):
            if root is None:
                return 0
            sumlist = [num + root.val for num in sumlist] + [root.val]
            return sumlist.count(sum) + dfs(root.left, sumlist) + dfs(root.right, sumlist)
        return dfs(root, [])


# 展开版，易理解
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:

        def dfs(root, sumlist):
            if root is None:
                return 0

            sumlist = [num+root.val for num in sumlist]
            sumlist.append(root.val)

            count = 0
            for num in sumlist:
                if num == sum:
                    count += 1
            # count = sumlist.count(sum)

            return count + dfs(root.left, sumlist) + dfs(root.right, sumlist)

        return dfs(root, [])


# ---------------------------------
if __name__ == "__main__":
    # test on LeetCode online.
    pass

File: C++/pathSum.cpp
/**
 * @File    : pathSum.cpp
 * @Brief   : 路径总和 III
 * @Link    : https://leetcode-cn.com/problems/path-sum-iii/
 * @Author  : Wei Li
 * @Date    : 2021-07-31
*/

#include <iostream>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    // 主函数
    int pathSum(TreeNode *root, int sum)
    {
        return root ? pathSumStartWithRoot(root, sum) +
                          pathSum(root->left, sum) + pathSum(root->right, sum)
                    : 0;
    }
    // 辅函数
    int pathSumStartWithRoot(TreeNode *root, int sum)
    {
        if (!root)
        {
            return 0;
        }
        int count = root->val == sum ? 1 : 0;
        count += pathSumStartWithRoot(root->left, sum - root->val);
        count += pathSumStartWithRoot(root->right, sum - root->val);
        return count;
    }
};

// ------------------------------
int main(int argc, char **argv)
{
    // test on LeetCode online.

    return 0;
}
